src/lib/tls/tls_callbacks.cpp

```diff
@@ -147,87 +143,123 @@ bool TLS::Callbacks::tls_verify_message(
    return verifier.verify_message(msg, sig);
    }
 
-std::pair<secure_vector<uint8_t>, std::vector<uint8_t>> TLS::Callbacks::tls_dh_agree(
-   const std::vector<uint8_t>& modulus,
-   const std::vector<uint8_t>& generator,
-   const std::vector<uint8_t>& peer_public_value,
-   const Policy& policy,
-   RandomNumberGenerator& rng)
+namespace {
+
+bool is_dh_group(const std::variant<TLS::Group_Params, DL_Group>& group)
    {
-   BigInt p = BigInt::decode(modulus);
-   BigInt g = BigInt::decode(generator);
-   BigInt Y = BigInt::decode(peer_public_value);
+   return std::holds_alternative<DL_Group>(group) ||
+          is_dh(std::get<TLS::Group_Params>(group));
+   }
 
-   /*
-    * A basic check for key validity. As we do not know q here we
-    * cannot check that Y is in the right subgroup. However since
-    * our key is ephemeral there does not seem to be any
-    * advantage to bogus keys anyway.
-    */
-   if(Y <= 1 || Y >= p - 1)
-      throw TLS_Exception(Alert::IllegalParameter,
-                          "Server sent bad DH key for DHE exchange");
+DL_Group get_dl_group(std::variant<TLS::Group_Params, DL_Group> group)
+   {
+   BOTAN_ASSERT_NOMSG(is_dh_group(group));
 
-   DL_Group group(p, g);
+   // TLS 1.2 allows specifying arbitrary DL_Group parameters in-lieu of
+   // a standardized DH group identifier. TLS 1.3 just offers pre-defined
+   // groups.
+   return std::visit(overloaded
+      {
+      [](DL_Group dl_group) { return dl_group; },
+      [&](TLS::Group_Params group_param) { return DL_Group(group_param_to_string(group_param)); }
+      }, std::move(group));
+   }
 
-   if(!group.verify_group(rng, false))
-      throw TLS_Exception(Alert::InsufficientSecurity,
-                          "DH group validation failed");
+}
 
-   DH_PublicKey peer_key(group, Y);
+std::unique_ptr<PK_Key_Agreement_Key> TLS::Callbacks::tls_generate_ephemeral_key(
+   const std::variant<TLS::Group_Params, DL_Group> group,
+   RandomNumberGenerator& rng)
+   {
+   if(is_dh_group(group))
```

Comment:
Just to check - there is something higher in the stack that is ensuring that in 1.3 we are only using group IDs and not explicit groups, right?

Reply:
TLS 1.3 doesn't support explicit statement of groups. The only place where this callback is called in the TLS 1.3 implementation is here: 

https://github.com/randombit/botan/pull/3322/files#diff-2cbf91ed19123a472976d7d1e316ab16b24e1a4062d6c97dfcb5af7d630e4d1fR44-R46

```
      Key_Share_Entry(const TLS::Group_Params group, Callbacks& cb, RandomNumberGenerator& rng)
         : m_group(group)
         , m_private_key(cb.tls_generate_ephemeral_key(group, rng))
```

... and we only have a group ID handy there.